force-app/main/default/classes/SOQL.cls

```diff
@@ -134,6 +165,24 @@ public with sharing class SOQL {
     //  The body of the method
     //}    
 
+        public static List<Lead> getLeadsBySourceOrTitle(String leadSource , String keyword){
+            if (String.isBlank(leadSource) && String.isBlank(keyword)) {
+                return new List<Lead>();
+            }
+
+            //SOQL conditions
+            List<String> conditions = new List<String>();
+            if(!String.isBlank(leadSource)){
+                conditions.add('LeadSource =:leadSource');
+            }
+            if(!String.isBlank(keyword)){
+                keyword = '%' + keyword + '%';
+                conditions.add('Title LIKE: keyword');
+            }
+            String query ='SELECT Id, LastName, Title, Company, LeadSource, Email, Phone FROM Lead WHERE ' + String.join(conditions, ' OR ');
+            return Database.query(query);
```

Comment:
Using dynamic SOQL here made this a lot more complicated than it needed to be.  It looks like you are trying to catch if one of the inputs is blank to not return on that condition. That is not in the requirements, but In that case you could check that they are not blank directly in your query. `(LeadSource = :leadSource AND LeadSource <> '') OR (Title LIKE :keyword AND Title <> '')`